src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -45,127 +53,135 @@ namespace Microsoft.VisualStudio.LanguageServices.Packaging
+        /// </summary>
+        private Task<ImmutableArray<PackageSource>>? _packageSourcesTask;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public PackageInstallerService(
             IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider,
             VisualStudioWorkspaceImpl workspace,
             SVsServiceProvider serviceProvider,
+            [Import("Microsoft.VisualStudio.Shell.Interop.SAsyncServiceProvider")] object asyncServiceProvider,
             IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
-            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
-            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2> packageInstaller,
-            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller> packageUninstaller,
-            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider> packageSourceProvider)
-            : base(threadingContext, workspace, SymbolSearchOptions.Enabled,
-                              SymbolSearchOptions.SuggestForTypesInReferenceAssemblies,
-                              SymbolSearchOptions.SuggestForTypesInNuGetPackages)
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices>? packageInstallerServices,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2>? packageInstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller>? packageUninstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider>? packageSourceProvider)
+            : base(threadingContext,
+                   workspace,
+                   SymbolSearchOptions.Enabled,
+                   SymbolSearchOptions.SuggestForTypesInReferenceAssemblies,
+                   SymbolSearchOptions.SuggestForTypesInNuGetPackages)
         {
             _workspace = workspace;
+
             _serviceProvider = serviceProvider;
+            // MEFv2 doesn't support type based contract for Import above and for this particular contract
+            // (SAsyncServiceProvider) actual type cast doesn't work. (https://github.com/microsoft/vs-mef/issues/138)
+            // workaround by getting the service as object and cast to actual interface
+            _asyncServiceProvider = (Shell.IAsyncServiceProvider)asyncServiceProvider;
+
             _editorAdaptersFactoryService = editorAdaptersFactoryService;
             _packageInstallerServices = packageInstallerServices;
             _packageInstaller = packageInstaller;
             _packageUninstaller = packageUninstaller;
             _packageSourceProvider = packageSourceProvider;
+
+            _workQueue = new AsyncBatchingWorkQueue<(bool solutionChanged, ProjectId? changedProject)>(
+                TimeSpan.FromSeconds(1),
+                this.ProcessWorkQueueAsync,
+                equalityComparer: null,
```

Comment:
Any reason here not to use the default equality comparer for the Tuple? If there's multiple project changes pushed it at the same time, there's no reason to have them both in the queue.

Reply:
good point.